From: Cliff Wickman <cpw@sgi.com>
Subject: x86, UV: Fix smp_processor_id() use in a preemptable region
Git-commit: 00b30cf04a775b5292ab704f782394e36e25617d
Patch-mainline: Yes
References: bnc#743212

Calling smp_processor_id() from within a preemptable region will issue
a warning if DEBUG_PREEMPT is set.

The calls to get_cpu()/put_cpu() bound the region where the return cpu
number is actually used and make it non-preemptable.  A DEBUG_PREEMPT
warning is prevented.
UV does not support cpu hotplug yet, but this is a step toward that ability.

Diffed against 3.0.0-rc3

Signed-off-by: Cliff Wickman <cpw@sgi.com>
Signed-off-by: Rafael J. Wysocki <rjw@suse.de>
---
 arch/x86/platform/uv/tlb_uv.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

Index: linux-3.0-SLE11-SP2/arch/x86/platform/uv/tlb_uv.c
===================================================================
--- linux-3.0-SLE11-SP2.orig/arch/x86/platform/uv/tlb_uv.c
+++ linux-3.0-SLE11-SP2/arch/x86/platform/uv/tlb_uv.c
@@ -1331,9 +1331,10 @@ static ssize_t tunables_write(struct fil
 
 	instr[count] = '\0';
 
-	bcp = &per_cpu(bau_control, smp_processor_id());
-
+	cpu = get_cpu();
+	bcp = &per_cpu(bau_control, cpu);
 	ret = parse_tunables_write(bcp, instr, count);
+	put_cpu();
 	if (ret)
 		return ret;
 
